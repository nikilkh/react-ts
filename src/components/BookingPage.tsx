import { useEffect, useState } from "react";
import { useLocation, useParams } from "react-router-dom";
import screen from "./screen.svg";
import whiteSeat from "./whiteSeat.svg";
import greySeat from "./greySeat.svg";
import blueSeat from "./blueSeat.svg";
import ConfirmBooking from "./ConfirmBooking";
import { AllSeatsStyle, BookingPageStyle, RowSeatStyle, ScreenStyle, SeatNumberStyle, SeatStyle, SittingSeatStyle } from "../styles/BookingPage.style";
type movie = {
  id: number
}
interface stateType {
  movie: movie
}



export default function BookingPage() {
  const IMG_URL = "https://image.tmdb.org/t/p/w500";

  const [selectedSeats, setSelectedSeats] = useState<string[] >([]);
  let [confirmedSeats, setConfirmedSeats] = useState<string[] | undefined | null >([]);
  const [modalVisible, setModalVisible] = useState<boolean>(false);
  const [bookingHomeVisible, setBookingHomeVisible] = useState<boolean>(true);
  let seatsForPriceCalc = [];
  const data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

  const seats = [
    {
      row: "A",
      seat: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    },
    {
      row: "B",
      seat: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    },
    {
      row: "C",
      seat: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    },
    {
      row: "D",
      seat: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    },
  ];
  let location= useLocation();
const state = location.state as stateType;
  // console.log(location.state.movie.id, location.state.movie.poster_path);
  function handleSeatClick(id: string) {
    if (selectedSeats?.includes(id)) {
      let newSelectedSeats = selectedSeats.filter((seat: string) => {
        return id !== seat;
      });
      setSelectedSeats(newSelectedSeats);
    } else if (selectedSeats.length < 10) {
      setSelectedSeats(() => [...selectedSeats, id]);
    } else alert("you can not select more than 10 seats");
  }

  function confirmBooking() {
    localStorage.setItem(state.movie.id.toString() , (selectedSeats.toString()));
    
    setConfirmedSeats(
      localStorage.getItem(state.movie.id.toString())?.split(",")
      );
    setModalVisible(true);
    setBookingHomeVisible(false);
  }

  function closeModal() {
    setModalVisible(false);
    setBookingHomeVisible(true);
  }

  useEffect(() => {
    if (localStorage.getItem(state.movie.id.toString()) !== null || undefined) {
      setSelectedSeats(
        localStorage.getItem(state.movie.id.toString()).split(",")
      );
      setConfirmedSeats(
        localStorage.getItem(state.movie.id.toString())?.split(",")
      );
      
    }
  }, []);

  // console.log(location.state.movie.id);
  return (
    <>
    `<BookingPageStyle
        style={
          bookingHomeVisible
            ? { background: "white" }
            : { background: "rgba(229, 229, 229, 0.6)", pointerEvents: "none" }
        }
      >
        <ScreenStyle>
          <img src={screen}></img>
        </ScreenStyle>

        <div style={{ margin: "auto" }}>
          <SeatNumberStyle>
            {data.map((num) => {
              return (
                <SeatStyle>
                  <h1>{num}</h1>
                </SeatStyle>
              );
            })}
          </SeatNumberStyle>
        </div>

        <AllSeatsStyle>
          {seats.map((item) => {
            return (
              <RowSeatStyle>
                <div>
                  <h1>{item.row}</h1>
                </div>
                {item.seat.map((seatNumber) => {
                  return (
                    <>
                      <SittingSeatStyle
                        onClick={() => {
                          handleSeatClick(item.row + seatNumber);
                        }}
                      >
                        {confirmedSeats?.includes(item.row + seatNumber) ? (
                          <img key={item.row + seatNumber} src={greySeat}></img>
                        ) : selectedSeats.includes(item.row + seatNumber) ? (
                          <img key={item.row + seatNumber} src={blueSeat}></img>
                        ) : (
                          <img
                            key={item.row + seatNumber}
                            src={whiteSeat}
                          ></img>
                        )}
                      </SittingSeatStyle>
                    </>
                  );
                })}
              </RowSeatStyle>
            );
          })}
        </AllSeatsStyle>

        <button onClick={confirmBooking}>Confirm Booking</button>
      </BookingPageStyle>

      {modalVisible ? (
        <ConfirmBooking closeModal={closeModal} selectedSeats={selectedSeats} confirmedSeats={confirmedSeats}>

        </ConfirmBooking>

        
      ) : null}
    </>
  );
}
