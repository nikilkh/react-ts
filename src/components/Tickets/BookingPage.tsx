import { useEffect, useState } from "react";
import { useLocation } from "react-router-dom";
import screen from "../SVGs/screen.svg";
import whiteSeat from "../SVGs/whiteSeat.svg";
import greySeat from "../SVGs/greySeat.svg";
import blueSeat from "../SVGs/blueSeat.svg";

import { ConfirmBooking } from "./ConfirmBooking";
import {
  AllSeatsStyle,
  BookingPageFlex,
  BookingPageStyle,
  LetterSeatStyle,
  RowSeatStyle,
  ScreenStyle,
  SeatNumberStyle,
  SeatStyle,
} from "./BookingPage.style";
import { GlobalStyle } from "../../styles/Global.styled";
import { Button } from "../Button/Button";
type movie = {
  id: number;
};
interface stateType {
  movie: movie;
}

type BookingProps = {
  children: React.ReactNode;
};

export default function BookingPage(props: BookingProps) {
  const [selectedSeats, setSelectedSeats] = useState<string[]>([]);
  const [confirmedSeats, setConfirmedSeats] = useState<string[]>([]);
  const [modalVisible, setModalVisible] = useState<boolean>(false);
  const [bookingHomeVisible, setBookingHomeVisible] = useState<boolean>(true);
  const data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

  const seats = {
    row: ["A", "B", "C", "D"],
    data: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
  };

  let location = useLocation();
  const state = location.state as stateType;

  function handleSeatClick(id: string) {
    if (selectedSeats?.includes(id)) {
      let newSelectedSeats = selectedSeats.filter((seat: string) => {
        return id !== seat;
      });
      setSelectedSeats(newSelectedSeats);
    } else if (selectedSeats.length < 10) {
      setSelectedSeats(() => [...selectedSeats, id]);
    }
  }

  function confirmBooking() {
    setConfirmedSeats((prevSeats) => [...prevSeats, ...selectedSeats]);
    localStorage.setItem(
      state.movie.id.toString(),
      [...confirmedSeats, ...selectedSeats].join(",")
    );
    if (selectedSeats.length > 0) {
      setModalVisible(true);
      setBookingHomeVisible(false);
    } else {
      alert("select minimum 1 seat");
    }
  }

  function closeModal() {
    setModalVisible(false);
    setBookingHomeVisible(true);
    setSelectedSeats([]);
  }

  useEffect(() => {
    if (localStorage.getItem(state.movie.id.toString())) {
      setConfirmedSeats(
        localStorage.getItem(state.movie.id.toString())?.split(",") as string[]
      );
    }
  }, []);

  return (
    <>
      <GlobalStyle />
      <BookingPageStyle bookingHomeVisible={bookingHomeVisible}>
        <BookingPageFlex>
          <ScreenStyle>
            <img src={screen} alt="screen this side"></img>
          </ScreenStyle>

          <SeatNumberStyle>
            {data.map((num) => {
              return (
                <SeatStyle>
                  <h1>{num}</h1>
                </SeatStyle>
              );
            })}
          </SeatNumberStyle>

          <AllSeatsStyle>
            {seats.row.map((item) => {
              return (
                <RowSeatStyle>
                  <LetterSeatStyle>
                    <h1>{item}</h1>
                  </LetterSeatStyle>

                  {seats.data.map((seatNumber) => {
                    return (
                      <>
                        {confirmedSeats?.includes(item + seatNumber) ? (
                          <img
                            key={item + seatNumber}
                            src={greySeat}
                            alt="confirmed seat"
                          ></img>
                        ) : selectedSeats.includes(item + seatNumber) ? (
                          <img
                            onClick={() => {
                              handleSeatClick(item + seatNumber);
                            }}
                            src={blueSeat}
                            alt="selected seat"
                          ></img>
                        ) : (
                          <img
                            onClick={() => {
                              handleSeatClick(item + seatNumber);
                            }}
                            src={whiteSeat}
                            alt="seat"
                          ></img>
                        )}
                      </>
                    );
                  })}
                </RowSeatStyle>
              );
            })}
          </AllSeatsStyle>

          <Button onclick={confirmBooking}>Confirm Booking</Button>
        </BookingPageFlex>
      </BookingPageStyle>
      {modalVisible ? (
        <ConfirmBooking
          closeModal={closeModal}
          selectedSeats={selectedSeats}
          confirmedSeats={confirmedSeats}
        >
          {props.children}
        </ConfirmBooking>
      ) : null}
    </>
  );
}
