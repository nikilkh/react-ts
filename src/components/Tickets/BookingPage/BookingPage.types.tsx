export type movie = {
    id: number;
  };
  
export interface stateType {
    movie: movie;
  }
  
export type BookingProps = {
    children: React.ReactNode;
  };