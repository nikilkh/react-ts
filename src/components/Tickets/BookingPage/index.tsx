export * from "./BookingPage";
export * from "./BookingPage.types";