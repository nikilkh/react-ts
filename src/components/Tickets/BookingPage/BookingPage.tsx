import { useEffect, useState } from "react";
import { useLocation } from "react-router-dom";
import { Screen } from "../../SVGs/Screen";

import { ConfirmBooking } from "../ConfirmBooking";
import {
  AllSeatsStyle,
  BookingPageFlex,
  BookingPageStyle,
  LetterSeatStyle,
  RowSeatStyle,
  ScreenStyle,
  SeatNumberStyle,
  SeatStyle,
} from "../BookingPageStyle/index";
import { GlobalStyle } from "../../../styles/Global.styled";
import { Button } from "../../Button/Button";
import { Seat } from "../../SVGs/Seat";
import { BookingProps, stateType } from "./BookingPage.types";

export const BookingPage = (props: BookingProps) => {
  const [selectedSeats, setSelectedSeats] = useState<string[]>([]);
  const [confirmedSeats, setConfirmedSeats] = useState<string[]>([]);
  const [modalVisible, setModalVisible] = useState<boolean>(false);
  const [bookingHomeVisible, setBookingHomeVisible] = useState<boolean>(true);

  const seats = {
    row: ["A", "B", "C", "D"],
    data: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
  };

  const location = useLocation();
  const state = location.state as stateType;

  function handleSeatClick(id: string) {
    if (selectedSeats?.includes(id)) {
      let newSelectedSeats = selectedSeats.filter((seat: string) => {
        return id !== seat;
      });
      setSelectedSeats(newSelectedSeats);
    } else if (selectedSeats.length < 10) {
      setSelectedSeats(() => [...selectedSeats, id]);
    }
  }

  function confirmBooking() {
    setConfirmedSeats((prevSeats) => [...prevSeats, ...selectedSeats]);
    localStorage.setItem(
      state.movie.id.toString(),
      [...confirmedSeats, ...selectedSeats].join(",")
    );
    if (selectedSeats.length > 0) {
      setModalVisible(true);
      setBookingHomeVisible(false);
    } else {
      alert("select minimum 1 seat");
    }
  }

  function closeModal() {
    setModalVisible(false);
    setBookingHomeVisible(true);
    setSelectedSeats([]);
  }

  useEffect(() => {
    if (localStorage.getItem(state.movie.id.toString())) {
      setConfirmedSeats(
        localStorage.getItem(state.movie.id.toString())?.split(",") as string[]
      );
    }
  }, []);

  return (
    <>
      <GlobalStyle />
      <BookingPageStyle bookingHomeVisible={bookingHomeVisible}>
        <BookingPageFlex>
          <ScreenStyle>
            <Screen />
          </ScreenStyle>

          <SeatNumberStyle>
            {seats.data.map((num) => {
              return (
                <SeatStyle>
                  <h1>{num}</h1>
                </SeatStyle>
              );
            })}
          </SeatNumberStyle>

          <AllSeatsStyle>
            {seats.row.map((item) => {
              return (
                <RowSeatStyle>
                  <LetterSeatStyle>
                    <h1>{item}</h1>
                  </LetterSeatStyle>

                  {seats.data.map((seatNumber) => {
                    return (
                      <>
                        {(() => {
                          switch (true) {
                            case confirmedSeats.includes(item + seatNumber):
                              return <Seat colorN="#626262" />;
                            case selectedSeats.includes(item + seatNumber):
                              return (
                                <Seat
                                  colorN="#724FD8"
                                  onClick={() => {
                                    handleSeatClick(item + seatNumber);
                                  }}
                                />
                              );
                            default:
                              return (
                                <Seat
                                  colorN="#DADADA"
                                  onClick={() => {
                                    handleSeatClick(item + seatNumber);
                                  }}
                                />
                              );
                          }
                        })()}
                      </>
                    );
                  })}
                </RowSeatStyle>
              );
            })}
          </AllSeatsStyle>

          <Button onclick={confirmBooking}>Confirm Booking</Button>
        </BookingPageFlex>
      </BookingPageStyle>
      {modalVisible ? (
        <ConfirmBooking
          closeModal={closeModal}
          selectedSeats={selectedSeats}
          confirmedSeats={confirmedSeats}
        >
          {props.children}
        </ConfirmBooking>
      ) : null}
    </>
  );
};
