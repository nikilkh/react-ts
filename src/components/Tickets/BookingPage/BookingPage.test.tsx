import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { BookingPage } from "./BookingPage";

describe("renders learn react link", () => {
  it("should render hello world", () => {
    const { container } = render(
      <BrowserRouter>
        <BookingPage children={undefined} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
