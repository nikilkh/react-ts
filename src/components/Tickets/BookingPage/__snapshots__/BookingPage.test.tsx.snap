// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link should render hello world 1`] = `<div />`;
